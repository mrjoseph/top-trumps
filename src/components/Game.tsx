import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux'
import SelectOption from './SelectOption';
import { Link } from 'react-router-dom';
import {checkAllCardsVisibleState ,getWinningPlayer} from '../utils';
import { Controls, PlayButton } from './game.styles';
import ActivePlayers from './Players';

interface IGame {
    pageNumber: number 
    numberOfPlayers: number
    activePlayers: any[]
    count: number
}

const INITIAL_STATE: IGame = {
    pageNumber: 1,
    numberOfPlayers: 2,
    activePlayers: [],
    count: 0
};

interface iGame {
    location: Location
}

interface Location {
    pathname: string
}

const Game = (props:iGame): JSX.Element => {
    const { location:{ pathname} } = props;
    const gameType = pathname.replace('/','');
    const dispatch = useDispatch();

    useEffect(() => {
        dispatch({ type: 'CLEAR_GAME' })
    },[dispatch]);

    const [state, setState] = useState<IGame>(INITIAL_STATE);
    const gameReducer = useSelector((state:any) => state.gameReducer)
    
    const pages = (gameType: string): number => {
        const pageNumber = (gameType === 'starships') ? 4 : 9;
       if(state.pageNumber >= pageNumber) return 1
        return state.pageNumber +1
    }

    const handleClick = (): void => {
        dispatch({ type: 'FETCH_REQUEST', 
        payload: {
            pageNumber: state.pageNumber,
            numberOfPlayers,
            gameType: gameType
        }
     })
        setState({
            ...state,  
            pageNumber: pages(gameType),
        })  

    }
    const handeChange = (players: number): void => {
        dispatch({ type: 'CLEAR_GAME' })
        setState({
            ...state, 
            numberOfPlayers:players
        })  
    }
    const { numberOfPlayers } = state;
    const { starwars, isFetched, error: {message}, activeCards, gameScore } = gameReducer;  
    
    console.log('gameScore', gameScore);
    useEffect(() => {
        const allCardsVisible = checkAllCardsVisibleState(activeCards);
        if(allCardsVisible) {
            console.log('dispatch')
            dispatch({ 
                type: 'GAME_SCORE',
                payload: { 
                    winningPlayers: getWinningPlayer(activeCards),
                    numberOfPlayers
                }
            }) 
        }
    },[dispatch, numberOfPlayers, activeCards]);
    
    return (
        <div>
            <Controls>
                <SelectOption handleChange={handeChange} count={state.numberOfPlayers} />
                <PlayButton onClick={handleClick}>PLAY</PlayButton>
            </Controls>
            
            { isFetched && starwars.length ?
            <ActivePlayers numberOfPlayers={numberOfPlayers} starwars={starwars} />: <div> Ready... </div> }
            {message && <p>Someting went wrong...</p>}
            <div><Link to="/">Home</Link></div>
            <div><Link to="/results">Results</Link></div>
        </div>
    )
};

export default Game;