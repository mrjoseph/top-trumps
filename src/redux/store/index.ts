import { combineReducers, createStore, applyMiddleware } from 'redux';
import { gameReducer } from '../reducers/reducers';
import createSagaMiddleware from 'redux-saga';
import mySaga from '../saga/sagas';

const sagaMiddleware = createSagaMiddleware()

const routeReducer = combineReducers({
    gameReducer
  });

export type AppState = ReturnType<typeof routeReducer>

const middleware = [sagaMiddleware];

const store = createStore(routeReducer, applyMiddleware(...middleware));
sagaMiddleware.run(mySaga);  
export default store;
