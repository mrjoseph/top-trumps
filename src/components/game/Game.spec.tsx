import React from 'react';
import Game from './Game';
import renderer from 'react-test-renderer';
import { Provider} from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import store from '../../redux/store';

describe('Game', () => {
    it('renders correctly', () => {
        interface IGame {
            location: Location
        }
        interface Location {
            pathname: string
        }

        const props:IGame = {
            location: {
                pathname: 'people/'
            }
        }
        const tree = renderer.create(
            <Router>
            <Provider store={store}>
                <Game {...props}/>
            </Provider>
            </Router>).toJSON();
        expect(tree).toMatchSnapshot();
      });
});