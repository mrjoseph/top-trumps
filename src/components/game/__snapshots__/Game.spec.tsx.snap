// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders correctly 1`] = `
<div>
  <div
    className="sc-ifAKCX khmXQv"
  >
    <label>
      Pick number of player:
      <select
        onChange={[Function]}
        value={2}
      >
        <option
          value="1"
        >
          1
        </option>
        <option
          value="2"
        >
          2
        </option>
        <option
          value="3"
        >
          3
        </option>
        <option
          value="4"
        >
          4
        </option>
      </select>
    </label>
    <button
      className="sc-EHOje dbWVjI"
      onClick={[Function]}
    >
      PLAY
    </button>
  </div>
  <div>
     Ready... 
  </div>
  <div>
    <a
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
  </div>
  <div>
    <a
      href="/results"
      onClick={[Function]}
    >
      Results
    </a>
  </div>
</div>
`;
